### DISCUSSION TITLE
next/headers: Next 13 Jest test fails using 'next/headers'

### DISCUSSION THREAD

**Discussion by njk112** [vercel/next.js#44270] [Help] [answered]
I am using Next.JS 13, and have a component which is using

```js
import { cookies } from 'next/headers';
```

The simplified component:

```js
import { cookies } from 'next/headers';

function Simple() {
  console.log({ cookies });
  return <div>Simple</div>;
}

export default Simple;
```

and the simplified test:

```js
import { render } from '@testing-library/react';

import Simple from './Simple';

describe('<Simple />', () => {
  it('should render Simple component', () => {
    render(<Simple />);
  });
});
```

When I have any reference to the `cookies`, in this case

```js
  console.log({ cookies });
```

I always get the following error when running jest tests:

```
     x NEXT_RSC_ERR_CLIENT_IMPORT: next/headers
       ,-[1:1]
     1 | import { cookies } from 'next/headers';
       : ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     2 | 
     3 | function Simple() {
     4 |   console.log({ cookies });
```

Any ideas how to fix it?

I have tried mocking the 'next/headers' with jest.mock, like so, but to no avail:

```js
jest.mock('next/headers', () => ({
  cookies: jest.fn(),
}));
```

My `jest.config.js`

```js
const nextJest = require('next/jest');

const createJestConfig = nextJest({
  dir: './',
});

const customJestConfig = {
  setupFilesAfterEnv: ['./.jest/setup.js'],
  testEnvironment: 'jest-environment-jsdom',
};

module.exports = async () => ({
  ...(await createJestConfig(customJestConfig)()),
});
```

My `./jest/setup.js`

```js
require('jest-fetch-mock').enableMocks();

import '@testing-library/jest-dom';
```

Package versions

```json
    "next": "13.1.0",
    "jest": "^29.3.1",
    "jest-environment-jsdom": "^29.3.1",
    "jest-fetch-mock": "^3.0.3",
    "@testing-library/jest-dom": "^5.16.5",
    "@testing-library/react": "^13.4.0",
```

**Comment by iSplasher** [comment]
Encountering this error too. At first glance, the error appears to be that NextJS thinks we're running on the client-side, but I even tried setting `"testEnvironment": "node"` with no success.

How are people testing their NextJS code?

**Reply by iSplasher** [reply]
Ok so apparently you're only allowed to import those Server Component functions in code under the `/app` directory.
I tested this by attempting to import them in an `pages/api` route and I got this error:
```
./src/pages/api/graphql.ts

You're importing a component that needs next/headers. That only works in a Server Component but one of its parents is marked with "use client", so it's a Client Component.

   ,-[2:1]
 2 | import Cors from 'cors';
 3 | import { GraphQLFormattedError } from 'graphql';
 4 | import morgan from 'morgan';
 5 | import { cookies } from 'next/headers';
   : ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 6 | 
 7 | import { ApolloServer } from '@apollo/server';
 8 | import { startServerAndCreateNextHandler } from '@as-integrations/next';
   `----
```

This seems rather shortsighted, I'd expect the error only to appear when the functions are called, not imported. How are you supposed to test files that import these functions?

**Reply by iSplasher** [reply]
My workaround for now is to pass these functions around, which is a PIA and not very elegant.

**Reply by hitesh0710** [reply]
@iSplasher Can you explain your solution how you passed these functions around?
Also if anyone can suggest better approach, I will really appreciate it. 
We need more support and docomentation on async server components.

**Reply by philwolstenholme** [reply]
I'd also be interesting to know a solution/workaround, I have this issue too

**Reply by peasandwell** [reply]
@iSplasher are you able to explain your workaround? It's impacting us and I'm curious to see if there's anything we haven't thought about

**Reply by colleen-love** [reply]
We ended up isolating the authentication code which references `cookies()` and mocking that in downstream unit tests. It'd be nice to be able to test that mission critical piece of authentication code, but it's good enough for now...

**Comment by red2678** [comment]
Needs attention imo. Not all my code lives in the app dir....nor should it.

**Comment by jordanmcdougall** [comment]
This discussion is related to open issues #47448 and #47299

**Comment by hnykda** [comment]
Got this as well. Does it, eh, means, that I essentially cannot test my server side functions without moving them to a totally different directory (+ cannot import/test anything that depends on what's in app dir)? I wanted to write my first test for my basic nextjs app, and hit this one 😢 ...

**Comment by bramarcade** [comment]
Confirming this also affects me.

**Comment by yusufseyrek** [comment]
I think the best option is to mock the server libraries like `next/header` atm. Any other solutions or does anyone know NextJS team is working on the issue?

**Comment by ryepup** [comment]
I found a workaround using [dynamic imports](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/import)

```diff
- import { cookies } from 'next/headers' 
+ const headers = import('next/headers')

- function foo() {
-   const c = cookies()
+ async function foo() {
+   const c = (await headers).cookies()
  // do something with cookies
}
```

With that in place, `jest.mock('next/headers')` works, and `next build` decides that pages/routes/layouts that call `foo` need to be rendered server-side, not pre-rendered.

**Reply by jordanmcdougall** [reply]
@ryepup Thank you, this seems to have gotten me a little further forward with my tests - but can you expand on how you are mocking `next/headers`?

Using just `jest.mock('next/headers')` results in the following error: `Invariant:  Method expects to have requestAsyncStorage, none available`

**Reply by ryepup** [reply]
@jordanmcdougall the mocking needed will depend on what you're testing.

For me, I have a really simple check I'm wanting to test, effectively

```ts
const c = (await headers).cookies()
return c.get('foo')?.value
```

To mock that, I'm using some setup code like:

```js
jest.mock('next/headers', () => {
  return { // <-- this object gets returned by `import('next/headers')`
    cookies: () => {
      return {
        get: jest.fn(),
      }
    },
  }
})
```
There are a lot of different ways to setup test mocks, depending on what you need https://jestjs.io/docs/manual-mocks

**Reply by nhollas-ben** [reply]
I get the same issue, did you ever manage to fix it?

`Invariant:  Method expects to have requestAsyncStorage, none available`

**Comment by njk112** [comment, marked as answer]
I have observed that the issue stems from the `next/jest` setup, where [`SWC`](https://nextjs.org/docs/pages/building-your-application/optimizing/testing#jest-and-react-testing-library) is used as the transform. Therefore made couple of changes:

Added new `jest.config.js` file with the following configuration:

```javascript
module.exports = {
	setupFilesAfterEnv: ["./jest/setup.js"],
	testEnvironment: "jest-environment-jsdom",
	transform: {
		"^.+\\.(js|jsx|ts|tsx)$": ["babel-jest", { presets: ["next/babel"] }],
	},
};
```

The purpose of this new configuration file is to replace SWC with Babel for transforming JS, JSX, TS, and TSX files.

Added `server-only.js` file to mock directory to simulate the server components. This mock prevents the error from being thrown.

```javascript
module.exports = {};
```

[Codesandbox](https://codesandbox.io/p/github/njk112/next-js-rsc-testing/main) or [github repo](https://github.com/njk112/next-js-rsc-testing) for example

**Comment by philwolstenholme** [comment]
https://github.com/vercel/next.js/pull/52393 should fix this in v13.4.10

**Comment by jpenna** [comment]
The fix was merged, but you have to include `/** @jest-environment node */` to the top of the file you are testing (check the PR #52393 ) or use the `projects` setting in `jest.config.js` ([docs](https://jestjs.io/docs/configuration#projects-arraystring--projectconfig)).

I've done this:

```js
const customJestConfig = {
  collectCoverage: true,
  coverageDirectory: 'coverage',
  coverageProvider: 'v8',
  projects: [
    {
      ...baseConfig,
      displayName: 'client',
      testMatch: ['<rootDir>/src/**/*.test.ts'],
      testEnvironment: 'jsdom',
    },
    {
      ...baseConfig,
      displayName: 'api',
      testMatch: ['<rootDir>/src/app/api/**/*.test.ts'],
      testEnvironment: 'node',
    },
  ],
};
```

**Reply by eric-burel** [reply]
The projects config doesn't seem to work with`next/jest`, or at least I couldn't make it work, see: https://github.com/vercel/next.js/discussions/59041#discussioncomment-12390982

**Comment by rgembalik** [comment]
I'm still getting this error during my tests:
```
NEXT_RSC_ERR_CLIENT_IMPORT: next/headers
```

The test file already has the annotation on the top:

```
/** @jest-environment node */
```

The version I've installed is `next@13.4.13`. Edit: also happens for `next@13.4.19`

Any ideas about what might be missing? Oh and if it changes anything, the test is for root layout, in which I am trying to recognize user session (via cookies) and pull user data based on that.

**Comment by Drag13** [comment]
Update to 13.5.4 fixed the issue without any config changes. 
Thanks guys.

**Reply by fbuireu** [reply]
`13.5.4` worked for me as well 😄

**Reply by eric-burel** [reply]
Hey folks, do you mean you manage to mock these values? I can't mock `next/headers` in v15, at least not the `headers()` function.

**Comment by tar-aldev** [comment]
Getting the same problem with cypress component testing. Updating next to 13.5.4 didn't help
